File: C++/strobogrammatic-number.cpp
// Time:  O(n)
// Space: O(1)

class Solution {
public:
    bool isStrobogrammatic(string num) {
        const int n = num.size();
        for (int i = 0; i <= n - 1 - i; ++i) {
            const auto it = lookup.find(num[n - 1 - i]);
            if (it == lookup.end() || num[i] != it->second) {
                return false;
            }
        }
        return true;
    }

private:
    const unordered_map<char, char> lookup{{'0', '0'}, {'1', '1'},
                                           {'6', '9'}, {'8', '8'}, 
                                           {'9', '6'}};
};

File: Python/strobogrammatic-number.py
# Time:  O(n)
# Space: O(1)

class Solution:
    lookup = {'0':'0', '1':'1', '6':'9', '8':'8', '9':'6'}

    # @param {string} num
    # @return {boolean}
    def isStrobogrammatic(self, num):
        n = len(num)
        i = 0
        while i <= n - 1 - i:
            if num[n - 1 - i] not in self.lookup or\
               num[i] != self.lookup[num[n - 1 - i]]:
                return False
            i += 1
        return True
